bee-crypto/src/ternary/sponge/batched_curlp/mod.rs

```diff
@@ -0,0 +1,197 @@
+    ///
+    /// It panics if the size of the batch exceeds `BATCH_SIZE`.
+    pub fn add(&mut self, input: TritBuf) {
+        assert!(self.trit_inputs.len() <= BATCH_SIZE);
+        self.trit_inputs.push(input);
+    }
+    /// Return the length of the current batch.
+    pub fn len(&self) -> usize {
+        self.trit_inputs.len()
+    }
+    /// Multiplex or interleave the input trits in the bash.
+    ///
+    /// Before doing the actual interleaving, each trit is encoded as two bits which are usually
+    /// refered as the low and high bit.
+    ///
+    /// | Trit | Low bit | High bit |
+    /// |------|---------|----------|
+    /// |  -1  |    1    |     0    |
+    /// |   0  |    0    |     1    |
+    /// |   1  |    1    |     1    |
+    ///
+    /// Then the low and high bits are interleaved into two vectors of integers. Each integer has
+    /// size `BATCH_SIZE` and there are `input_length` integers in each vector.  This means that
+    /// the low and high bits of the transaction number `N` in the batch are stored in the position
+    /// `N` of each integer.
+    ///
+    /// This step works correctly even if there are less than `BATCH_SIZE` inputs.
+    fn mux(&mut self) {
+        let count = self.trit_inputs.len();
+        for i in 0..self.bct_inputs.len() {
+            let bc_trit = self.bct_inputs.get_mut(i);
+
+            for j in 0..count {
+                match self.trit_inputs[j][i] {
+                    Btrit::NegOne => *bc_trit.lo |= 1 << j,
+                    Btrit::PlusOne => *bc_trit.hi |= 1 << j,
+                    Btrit::Zero => {
+                        *bc_trit.lo |= 1 << j;
+                        *bc_trit.hi |= 1 << j;
+                    }
+                }
+            }
+        }
+    }
+    /// Demultiplex the bits of the output to obtain the hash of the input with an specific index.
+    ///
+    /// This is the inverse of the `mux` function, but it is applied over the vector with the
+    /// binary encoding of the output hashes. Each pair of low and high bits in the `bct_hashes`
+    /// field is decoded into a trit using the same convention as the `mux` step with an additional
+    /// rule for the `(0, 0)` pair of bits which is mapped to the `0` trit.
+    fn demux(&self, index: usize) -> TritBuf {
+        let length = self.bct_hashes.len();
+        let mut result = Vec::with_capacity(length);
+
+        for i in 0..length {
+            let low = (self.bct_hashes.lo()[i] >> index) & 1;
+            // This is safe as the previous access to `self.lo` took care of checking the index.
+            let hi = (unsafe { *self.bct_hashes.hi().get_unchecked(i) } >> index) & 1;
+
+            let trit = match (low, hi) {
+                (1, 0) => Btrit::NegOne,
+                (0, 1) => Btrit::PlusOne,
+                (1, 1) => Btrit::Zero,
+                // This can only be `(0, 0)`.
+                _ => Btrit::Zero,
+            };
+
+            result.push(trit);
```

Comment:
Maybe controversial but there is some logic in `push` that is not needed since we already set the size.

Reply:
highly controversial! I like it.

Reply:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_mut_ptr :trollface:

Reply:
I did the change but after running the benchmarks I couldn't find any significant speedup so I think it's not worth it.

Reply:
Fair enough